api/v1alpha4/clusterclass_webhook.go

```diff
@@ -92,7 +96,32 @@ func (in *ClusterClass) validate(old *ClusterClass) error {
 	var allErrs field.ErrorList
 
 	// ensure all the references are within the same namespace
-	if in.Spec.Infrastructure.Ref != nil && in.Spec.Infrastructure.Ref.Namespace != in.Namespace {
+	allErrs = append(allErrs, in.validateRefNamespace()...)
+
+	// Ensure all references are templates
+	allErrs = append(allErrs, in.validateRefTemplates()...)
+
+	// Ensure that all the refs have valid apiVersions
+	allErrs = append(allErrs, in.validateRefAPIVersions()...)
+
+	// Ensure all machine deployments are unique
+	// for update:
+	//   also ensure that the old machine deployments still exist and new deployments are unique
+	allErrs = append(allErrs, in.validateMachineDeployments(old)...)
+
+	// Ensure sure no additional changes were applied.
+	allErrs = append(allErrs, in.validateAdditionalChanges(old)...)
+
+	if len(allErrs) > 0 {
+		return apierrors.NewInvalid(GroupVersion.WithKind("ClusterClass").GroupKind(), in.Name, allErrs)
+	}
+	return nil
+}
+
+func (in *ClusterClass) validateRefNamespace() field.ErrorList {
```

Comment:
Yeah, I think it would make sense to group `validateRefNamespace`, `validateRefTemplates` and `validateRefAPIVersions` under a `validateRef` function. Will probably have to adopt a similar pattern as https://github.com/kubernetes-sigs/cluster-api/pull/4904#discussion_r672673557 to have pathPrefix.